src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -369,11 +503,77 @@ await foreach (ConfigurationSetting setting in _client.GetConfigurationSettingsA
 
-        private async Task LoadKeyValuesRegisteredForRefresh(IDictionary<string, ConfigurationSetting> data, CancellationToken cancellationToken)
+        private async Task<Dictionary<string, ConfigurationSetting>> LoadAll(ConfigurationClient client, CancellationToken cancellationToken)
+        {
+            var serverData = new Dictionary<string, ConfigurationSetting>(StringComparer.OrdinalIgnoreCase);
+
+            // Use default query if there are no key-values specified for use other than the feature flags
+            bool useDefaultQuery = !_options.KeyValueSelectors.Any(selector => !selector.KeyFilter.StartsWith(FeatureManagementConstants.FeatureFlagMarker));
+
+            if (useDefaultQuery)
+            {
+                // Load all key-values with the null label.
+                var selector = new SettingSelector
+                {
+                    KeyFilter = KeyFilter.Any,
+                    LabelFilter = LabelFilter.Null
+                };
+
+                await CallWithRequestTracing(async () =>
+                {
+                    await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAsync(selector, cancellationToken).ConfigureAwait(false))
+                    {
+                        serverData[setting.Key] = setting;
+                    }
+                }).ConfigureAwait(false);
+            }
+
+            foreach (var loadOption in _options.KeyValueSelectors)
+            {
+                if ((useDefaultQuery && LabelFilter.Null.Equals(loadOption.LabelFilter)) ||
+                    _options.KeyValueSelectors.Any(s => s != loadOption &&
+                       string.Equals(s.KeyFilter, KeyFilter.Any) &&
+                       string.Equals(s.LabelFilter, loadOption.LabelFilter)))
+                {
+                    // This selection was already encapsulated by a wildcard query
+                    // Or would select kvs obtained by a different selector
+                    // We skip it to prevent unnecessary requests
+                    continue;
+                }
+
+                var selector = new SettingSelector
+                {
+                    KeyFilter = loadOption.KeyFilter,
+                    LabelFilter = loadOption.LabelFilter
+                };
+
+                await CallWithRequestTracing(async () =>
+                {
+                    await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAsync(selector, cancellationToken).ConfigureAwait(false))
+                    {
+                        serverData[setting.Key] = setting;
+                    }
+                }).ConfigureAwait(false);
+            }
+
+            // Block current thread for the initial load of key-values registered for refresh that are not already loaded
+            await Task.Run(() => LoadKeyValuesRegisteredForRefresh(client, serverData, cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult()).ConfigureAwait(false);

```

Comment:
Not related to this PR, but this looks not good. Looks like it should just be
`await LoadKeyValuesRegisteredForRefresh(client, serverData, cancellationToken).ConfigureAwait(false);`

Otherwise, like the comment says we're blocking a thread

Reply:
From the comment, it seems like it was intended to block the thread while loading key-values initially.
Otherwise what you suggested makes sense!

Reply:
I think the intention was to block the thread during startup. But we are also blocking the thread when `LoadAll` is invoked during `Load()`. So we may not need to block it here again.

Reply:
Thanks for the clarification.
I removed the blocking mechanism.